source/geod24/concurrency.d

```diff
@@ -1497,6 +1532,181 @@ private:
     assert(x[0] == 5);
 }
 
+/***********************************************************************
+
+    A common interface for objects that can be used in `select()`
+
+***********************************************************************/
+
+public interface Selectable
+{
+    /***********************************************************************
+
+        Try to read/write to the `Selectable` without blocking. If the
+        operation would block, queue and link it with the `sel_state`
+
+        Params:
+            ptr = pointer to the data for the select operation
+            sel_state = SelectState instace of the select call being executed
+            sel_id = id of the select call being executed
+
+    ***********************************************************************/
+
+    void selectWrite (void* ptr, SelectState sel_state, int sel_id);
+
+    /// Ditto
+    void selectRead (void* ptr, SelectState sel_state, int sel_id);
+}
+
+/***********************************************************************
+
+    An aggregate to hold neccessary information for a select operation
+
+***********************************************************************/
+
+public struct SelectEntry
+{
+    /// Reference to a Selectable object
+    Selectable selectable;
+
+    /// Pointer to the select data
+    void* select_data;
+
+    /***********************************************************************
+
+        Default ctor
+
+        Params:
+            selectable = A selectable interface reference
+            select_data = pointer to the data for the select operation
+
+    ***********************************************************************/
+
+    this (Selectable selectable, void* select_data)
+    {
+        this.selectable = selectable;
+        this.select_data = select_data;
+    }
+}
+
+/// Consists of the id and result of the select operation that was completed
+public alias SelectReturn = Tuple!(bool, "success", int, "id");
+
+/***********************************************************************
+
+    Block on multiple `Channel`s
+
+    Only one operation is completed per `select()` call
+
+    Params:
+        read_list = List of `Channel`s to select for read operation
+        write_list = List of `Channel`s to select for write operation
+
+    Return:
+        Returns success/failure status of the operation and the index
+        of the `Channel` that the operation was carried on. The index is
+        the position of the SelectEntry in `read_list ~ write_list`, ie
+        concatenated lists.
+
+***********************************************************************/
+
+public SelectReturn select (ref SelectEntry[] read_list, ref SelectEntry[] write_list)
```

Comment:
Notes for the future:
Having a `randomShuffle` here is a bit meh. Having mutation on the parameter for the benefit of the function is a red flag. But let's go ahead for now.

Reply:
`randomCover` could work?

Reply:
`randomCover` wont work. `SelectEntry` s get ID'ed in the order they are traversed.